### ISSUE TITLE
Add an image to post

### ISSUE THREAD

**Issue by MrAntonS** [hjiangsu/thunder#237]
Right now the only way to upload an image is to upload it through the browser and copy link to the image, I want to add an image field to post menu. Please let me know if I'm missing something, if not please assign me to this issue

**Comment by MrAntonS** [comment]
Different modes sound good, to allow images inside the post i would make another button in the text edit to attach the image. Other than that, I'm thinking about adding a carousel to the image_view, and add all images in there, and when you open the post switch back to normal mode and images in the text on places they should be on, what do you think?

**Comment by MrAntonS** [comment]
On the other hand I would just add attach image field and make it optional instead of modes, it makes it a little complicated for the user in my opinion.

**Comment by hjiangsu** [comment]
I think what we can do for an MVP is to start off with just the button to attach an image. Then, once we have that, we can slowly work towards adding the different modes as I suggested!

> Different modes sound good, to allow images inside the post i would make another button in the text edit to attach the image. Other than that, I'm thinking about adding a carousel to the image_view, and add all images in there, and when you open the post switch back to normal mode and images in the text on places they should be on, what do you think?

Go for it!

**Comment by MrAntonS** [comment]
Hmm, I was doing my research and find out couple things. To begin, I am using lemmy.fmhy.ml instance. When I was uploading images on it, looking at the requests it actually went through https://api.imgbb.com/1/upload?key=0551e2b3b8cabbc104781133018278f2 instead of /pictrs/image on the instance itself, it seems like they are using it as a bridge to the API and do not allow direct interactions. On the lemmy.world instance I created just for testing, they're directly communicate with an API without any middle steps. So my question is - is there a way to determine if instance has a bridge or no. We could try use a bridge on every instance, but I feel like there should be a way to upload images to every instance identically. I will check out other clients and see how they do it, but if you have any suggestion feel free to share them. Thank you in advance!

**Comment by MrAntonS** [comment]
I wrote the UI part, the only thing that is left it an API call for the image. I have a question, in your code you're using `community_bloc.dart` to process all API requests, I was thinking about adding ImageUpload event to that file, but, I can't wrap my head around the emit return and how to process it in the `create_post_page.dart`. Instead, I'm thinking about just implementing the function directly in `_CreatePostPageState` class. But that would add complexity to the code. Please let me know what do you think.

**Comment by hjiangsu** [comment]
I'll try to answer your question to the best of my abilities:

`community_bloc.dart` should generally contain all the logic for handling things related to data flow for events that happen within the `CommunityPage` view. I think we should stick to that convention and create a new event on there.

Here's a really quick and untested draft of something you could try:

1. Create a new event in `community_bloc` for the image uploading API

```dart
// community_event.dart
class UploadImageEvent extends CommunityEvent {
  final ByteData imageByteData; // I'm not sure what the format of the image being uploaded is, so just going to use a generic type here - this will change depending on how its implemented

  const UploadImageEvent({required this.imageByteData});
}
```

2. Create the logic for the associated event in `community_bloc`
```dart
// community_bloc.dart
Future<void> _uploadImageEvent(UploadImageEvent event, Emitter<CommunityState> emit) async {
    emit(state.copyWith(status: CommunityStatus.refreshing, communityId: state.communityId, listingType: state.listingType));
    
    // Perform logic to upload image here

    // Generate a random string token for the uploaded image (this is to allow us to determine a state change for an image upload)
    String imageUploadToken = "abcdefg";

    emit(state.copyWith(status: CommunityStatus.success, communityId: state.communityId, listingType: state.listingType, imageUploadToken: imageUploadToken));
  }
```

3. Add in any extra state variables to `community_bloc`
```dart
// community_state.dart
const CommunityState({
...,
this.imageUploadToken,
});

// follow the rest of the logic to allow for assigning imageUploadToken
```

3. Have a button that triggers a new event to happen in `community_bloc`

```dart
IconButton(
  onPressed: () {
      context.read<CommunityBloc>().add(UploadImageEvent(byteData: imageByteData));
  },
  icon: const Icon(
    Icons.upload,
    semanticLabel: 'Upload Image,
  ),
),
```

5. Create a listener inside of the `CreatePostPage` which listens to some specific event that is emitted from the `community_bloc` and update the logic inside `CreatePostPage`

```dart
Scaffold(
  body: BlocListener<CommunityBloc, CommunityState>(
      listenWhen: (previous, current) => (previous.status == CommunityStatus.refreshing && current.status == CommunityStatus.success && previous.imageUploadToken != current.imageUploadToken),
      listener: (context, state) {
        // This only triggers when listenWhen returns true
        setState(() => hasImageBeenUploaded = true)
      },
      child: Container(), // The rest of the logic for UI here
),
```

Hopefully this helps you wrap your head around how the logic is structured! Let me know if you have any other questions

TLDR:
- send an event to community_bloc within the create post page to trigger action
- the community_bloc "emits" a state when there is a change
- have the create post page listen for a specific event to occur
- update create post page UI based on the changes from the event

**Comment by hjiangsu** [comment]
For an fuller example, this is the create comment modal https://github.com/hjiangsu/thunder/blob/main/lib/post/widgets/create_comment_modal.dart. 

In here, I'm listening to the status of whether the comment has actually been successful before closing the modal:

This logic is a bit more complicated since its listening to two different blocs, depending on which is is available, and performs different actions based which one is available, but hopefully you get a general idea of what's happening

https://github.com/hjiangsu/thunder/blob/fd5d897cfc2d11d04a0eb4bfacb576744e03eb3f/lib/post/widgets/create_comment_modal.dart#L84-L107

**Comment by hjiangsu** [comment]
@MrAntonS If you'd like, you can create a PR just for the UI portion of this, and I can connect it with the API if that makes things easier too!

**Comment by MrAntonS** [comment]
No it's ok, I was a little busy for two days, I found your comment really helpful and will do this today, I've already implemented API part the only thing left is to create custom bloc and add event in there

**Comment by hjiangsu** [comment]
Alright no worries! Thanks for the update :D

**Comment by MrAntonS** [comment]
#305
